### PR TITLE
gh-104874: Document NewType.__supertype__

### LINKED ISSUE #104874: Document `NewType.__supertype__`
`typing.NewType` has always had an attribute `__supertype__` pointing to its supertype. It is widely used in the wild (e.g. https://github.com/pydantic/pydantic/blob/00158878cb1c6218cd58ee62248db56c65e4172c/pydantic/_internal/_typing_extra.py#L153, https://github.com/quora/pyanalyze/blob/bd7f520adc2d8b098be657dfa514d1433bea3b0c/pyanalyze/annotations.py#L456), but not documented. Let's document it and commit to leaving it stable.

### Linked PRs
* gh-104875
* gh-104906
* gh-104907

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Fixes #104874

* Issue: gh-104874

----
:books: Documentation preview :books:: https://cpython-previews--104875.org.readthedocs.build/

### CODE REVIEW

**Review by AlexWaygood — COMMENTED** [review]
Looks good. We also test for `__qualname__` and `__module__`; should we document those as well?

https://github.com/python/cpython/blob/c90a862cdcf55dc1753c6466e5fa4a467a13ae24/Lib/test/test_typing.py#L6658-L6669

**Comment by JelleZijlstra** [comment]
`__qualname__` is actually wrong as implemented (it's always the same as `__name__`) and `__module__` is mostly an implementation detail for pickling, so I'd prefer to leave those undocumented.

**Comment by AlexWaygood** [comment]
> `__module__` is mostly an implementation detail for pickling

I feel like it could be useful to be able to know the module the newtype was defined in, and it seems useful to mention the ways in which instances of `NewType` are similar to classes, even though they're not classes.

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -1648,6 +1648,18 @@ These are not used in annotations. They are building blocks for declaring types.
       UserId = NewType('UserId', int)
       first_user = UserId(1)
 
+   .. attribute:: __module__
+
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion

```

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -1648,6 +1648,17 @@ These are not used in annotations. They are building blocks for declaring types.
       UserId = NewType('UserId', int)
       first_user = UserId(1)
 
+   .. attribute:: __module__
+      The module in which the new type is defined.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   .. attribute:: __module__

      The module in which the new type is defined.
```

**Pull request merged**